arrow/src/main/java/org/apache/iceberg/arrow/DictEncodedArrowConverter.java

```diff
@@ -0,0 +1,225 @@
+        vector,
+        vectorHolder,
+        idx -> vector.set(idx, (BigDecimal) accessor.getDecimal(idx, precision, scale)));
+    return vector;
+  }
+
+  private static TimeStampVector toTimestampVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    TimeStampVector vector;
+    if (((Types.TimestampType) vectorHolder.icebergType()).shouldAdjustToUTC()) {
+      vector =
+          new TimeStampMicroTZVector(
+              vectorHolder.vector().getName(),
+              ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+              vectorHolder.vector().getAllocator());
+    } else {
+      vector =
+          new TimeStampMicroVector(
+              vectorHolder.vector().getName(),
+              ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+              vectorHolder.vector().getAllocator());
+    }
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getLong(idx)));
+    return vector;
+  }
+
+  private static BigIntVector toBigIntVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    BigIntVector vector =
+        new BigIntVector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getLong(idx)));
+    return vector;
+  }
+
+  private static Float4Vector toFloat4Vector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    Float4Vector vector =
+        new Float4Vector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getFloat(idx)));
+    return vector;
+  }
+
+  private static Float8Vector toFloat8Vector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    Float8Vector vector =
+        new Float8Vector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(vector, vectorHolder, idx -> vector.set(idx, accessor.getDouble(idx)));
+    return vector;
+  }
+
+  private static VarCharVector toVarCharVector(
+      VectorHolder vectorHolder, ArrowVectorAccessor<?, String, ?, ?> accessor) {
+    VarCharVector vector =
+        new VarCharVector(
+            vectorHolder.vector().getName(),
+            ArrowSchemaUtil.convert(vectorHolder.icebergField()).getFieldType(),
+            vectorHolder.vector().getAllocator());
+
+    initVector(
+        vector,
+        vectorHolder,
+        idx -> vector.set(idx, accessor.getUTF8String(idx).getBytes(StandardCharsets.UTF_8)));
```

Comment:
@nastra,

We need to use setSafe() method instead of set() method. As we are not setting the size initially, there is chance of getting IndexOutOfBoundsException if the varchar column value is lengthy.

Reply:
ah yes you're right, we need to use `setSafe(..)` here. I've updated the code.